File: server/voice_events.py
"""Send a voice call's page moves and live captions to the visitor's browser.

Retell's v3 web calls, the only kind left after 2026-09-30, no longer pass the
`metadata` event or live transcript updates through to the browser. Page
navigation ran on the first and the call panel's captions on the second. So the
server sends both itself, over Pusher Channels:

1. The browser makes up a random UUID, subscribes to the public channel
   `voice-<uuid>`, and puts the UUID in the call's metadata as `events_channel`
   when it creates the call.
2. Retell hands that metadata to this server in the `call_details` event.
3. For the rest of the call, main.py publishes each navigation event and the
   last few transcript lines to that channel.

A public channel is enough. Its name is a random UUID other visitors can't
guess, and no Pusher auth endpoint means one less thing to deploy. The id does
reach this call's trace metadata and Retell's call record, which only the
owner can read. PUSHER_SECRET is what really keeps calls private: with it,
anyone could list the open `voice-` channels and read or forge their events.
Unlike the app key, it must stay secret.

Publishing never holds up the voice reply. Events go out on background tasks,
one at a time so they arrive in order, and a failed send is logged and dropped:
a page that stops following along is better than an agent that goes quiet.
"""

import asyncio
import functools
import json
import os
import re
from concurrent.futures import ThreadPoolExecutor

import certifi
import pusher
from pusher.requests import RequestsBackend

__all__ = ["VoiceEvents", "caption_window", "channel_for"]

# Not secrets: the browser needs the key and cluster to subscribe, and the app
# id means nothing without the secret. Overridable per environment.
DEFAULT_APP_ID = "2196222"
DEFAULT_KEY = "3b1aa13ad53a1a55ff03"
DEFAULT_CLUSTER = "us3"

CHANNEL_PREFIX = "voice-"
# What crypto.randomUUID() produces. The id arrives in call metadata the browser
# chose, so anything else is refused rather than passed to Pusher.
_CHANNEL_ID = re.compile(r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}")

# Retell's update events carry the whole transcript; only the last few lines
# still change. Each line goes out with its index in Retell's transcript, and
# the browser replaces lines by index (client/src/lib/transcript.ts
# mergeVoiceLines). Guessing how a window lines up with what's on screen broke
# as soon as two lines changed between sends, which the throttle makes common.
TRANSCRIPT_WINDOW = 5
# Pusher rejects an event whose data is over 10 KB. Stay well under it.
MAX_EVENT_CHARS = 9000
MAX_LINE_CHARS = 4000
# Retell sends an update for nearly every word. Captions go out at most this
# often; the newest window always wins, so none of the text is lost.
CAPTION_INTERVAL_S = 0.25
# How long the end of a call waits for events still being sent.
CLOSE_TIMEOUT_S = 2.0
PUSHER_TIMEOUT_S = 3

# The pusher library blocks, so sends run on threads. Its own small pool: on
# the default one, a stalled Pusher could hold every worker and queue the DNS
# lookups the OpenAI and Pinecone clients make there.
_POOL = ThreadPoolExecutor(max_workers=4, thread_name_prefix="pusher")

_client: pusher.Pusher | None = None
_warned_unconfigured = False
_build_failed = False


class _CertifiBackend(RequestsBackend):
    """The library's requests backend, checking TLS against certifi.

    Left alone, it checks against a root list bundled with the library in
    2021, which neither a certifi update nor the Docker image can refresh.
    """

    def __init__(self, client, **options):
        super().__init__(client, **options)
        if client.ssl:
            self.options["verify"] = certifi.where()


def _pusher_client() -> pusher.Pusher | None:
    """The process-wide Pusher client, or None when PUSHER_SECRET is unset.

    Built on first use, not at import: main.py loads .env after importing its
    modules. One client for the whole process, so every call reuses its
    connection pool and a navigation event doesn't pay for a TLS handshake.
    """
    global _client, _warned_unconfigured, _build_failed
    if _client is not None:
        return _client
    if _build_failed:
        return None
    secret = os.getenv("PUSHER_SECRET")
    if not secret:
        if not _warned_unconfigured:
            _warned_unconfigured = True
            print(
                "[voice-events] PUSHER_SECRET is not set; voice calls won't move "
                "the page or show captions",
                flush=True,
            )
        return None
    try:
        _client = pusher.Pusher(
            app_id=os.getenv("PUSHER_APP_ID") or DEFAULT_APP_ID,
            key=os.getenv("PUSHER_KEY") or DEFAULT_KEY,
            secret=secret,
            cluster=os.getenv("PUSHER_CLUSTER") or DEFAULT_CLUSTER,
            ssl=True,
            timeout=PUSHER_TIMEOUT_S,
            backend=_CertifiBackend,
        )
    except Exception as e:  # noqa: BLE001 - a bad override must not stop the call
        # Runs inside the call_details handler, before the greeting is sent.
        # A config error won't fix itself, so don't retry and log it per call.
        _build_failed = True
        print(f"[voice-events] Pusher client not built: {type(e).__name__}: {e}", flush=True)
        return None
    return _client


def channel_for(call: object) -> str | None:
    """The Pusher channel a call's browser listens on, or None if it named none."""
    metadata = call.get("metadata") if isinstance(call, dict) else None
    raw = metadata.get("events_channel") if isinstance(metadata, dict) else None
    if not isinstance(raw, str) or not _CHANNEL_ID.fullmatch(raw):
        return None
    return CHANNEL_PREFIX + raw


def _chars(data: object) -> int:
    # The size of what _send hands to pusher: ASCII-only JSON, so characters
    # are bytes, which is what Pusher's 10 KB limit counts. (The library's own
    # check, sys.getsizeof under 30720, then passes with room to spare.)
    return len(json.dumps(data))


def caption_window(transcript: object) -> list[dict]:
    """The last few spoken lines, as {index, role, content}, small enough for
    one event.

    `index` is the line's position in Retell's transcript, which stays put as
    the line grows or is corrected, so the browser can replace it in place.
    Drops Retell's per-word timings and anything that isn't a visitor or agent
    line with text in it (their indices are simply skipped).
    """
    if not isinstance(transcript, list):
        return []
    lines = [
        {"index": i, "role": u["role"], "content": u["content"][:MAX_LINE_CHARS]}
        for i, u in enumerate(transcript)
        if isinstance(u, dict)
        and u.get("role") in ("agent", "user")
        and isinstance(u.get("content"), str)
        and u["content"].strip()
    ]
    window = lines[-TRANSCRIPT_WINDOW:]
    while len(window) > 1 and _chars({"transcript": window}) > MAX_EVENT_CHARS:
        window = window[1:]
    # One line can still be too big: ASCII-only JSON spells each non-ASCII
    # character as a six-character escape (twelve for an emoji). Cut it from
    # the end, so the start the browser already shows stays put.
    while window and _chars({"transcript": window}) > MAX_EVENT_CHARS:
        line = window[0]
        content = line["content"][: len(line["content"]) // 2]
        window = [{**line, "content": content}] if content.strip() else []
    return window


class VoiceEvents:
    """One voice call's outbox to its browser."""

    def __init__(self, channel: str, client: pusher.Pusher):
        self.channel = channel
        self._client = client
        # One send at a time. asyncio.Lock wakes waiters in the order they
        # queued, so events reach the browser in the order they were made.
        self._lock = asyncio.Lock()
        self._tasks: set[asyncio.Task] = set()
        self._pending_caption: list[dict] | None = None
        self._last_caption: list[dict] | None = None
        self._caption_task: asyncio.Task | None = None
        self._closed = False

    @classmethod
    def for_call(cls, call: object) -> "VoiceEvents | None":
        """An outbox for this call, or None if it has nowhere to send to."""
        channel = channel_for(call)
        if channel is None:
            return None
        client = _pusher_client()
        if client is None:
            return None
        return cls(channel, client)

    def navigation(self, meta: dict) -> None:
        """Move the visitor's page. Returns at once; the send happens later."""
        if self._closed:
            return
        self._spawn(self._send("navigation", meta))

    def transcript(self, transcript: object) -> None:
        """Show the call's latest lines. Updates that change nothing are skipped."""
        if self._closed:
            return
        window = caption_window(transcript)
        # Retell also sends an update when the speaker changes, with the same
        # text as the last one.
        if not window or window == (self._pending_caption or self._last_caption):
            return
        self._pending_caption = window
        if self._caption_task is None or self._caption_task.done():
            self._caption_task = self._spawn(self._drain_captions())

    async def aclose(self) -> None:
        """Stop taking events, and give the ones in flight a moment to go out."""
        self._closed = True
        pending = [t for t in self._tasks if not t.done()]
        if not pending:
            return
        _, late = await asyncio.wait(pending, timeout=CLOSE_TIMEOUT_S)
        for task in late:
            task.cancel()
        await asyncio.gather(*late, return_exceptions=True)

    async def _drain_captions(self) -> None:
        while self._pending_caption is not None:
            window, self._pending_caption = self._pending_caption, None
            # Only a delivered window counts as sent: after a failure, the next
            # update with the same text must go out, not be skipped as a repeat.
            if await self._send("transcript", {"transcript": window}):
                self._last_caption = window
            await asyncio.sleep(CAPTION_INTERVAL_S)

    def _spawn(self, coro) -> asyncio.Task:
        task = asyncio.create_task(coro)
        self._tasks.add(task)
        task.add_done_callback(self._tasks.discard)
        return task

    async def _send(self, event: str, data: dict) -> bool:
        """Publish one event; True if Pusher accepted it."""
        # Serialised here, ASCII-only. Given a dict, the library keeps
        # non-ASCII text raw and sizes it with sys.getsizeof, which counts an
        # emoji-bearing string at four bytes a character and rejects captions
        # that fit Pusher's limit. A string passes through unchanged, and the
        # browser decodes the same object either way.
        payload = json.dumps(data)
        async with self._lock:
            try:
                await asyncio.get_running_loop().run_in_executor(
                    _POOL, functools.partial(self._client.trigger, self.channel, event, payload)
                )
            except Exception as e:  # noqa: BLE001 - a lost event must not end the call
                print(f"[voice-events] {event} not sent: {type(e).__name__}: {e}", flush=True)
                return False
            return True

File: client/src/lib/voice-events.ts
import type { CreateWebCallRequest } from "@/types/api";
import type { VoiceLine } from "./transcript";
import type { NavigationMeta } from "./voice-bus";

/** Where a voice call's page moves and captions come from.
 *
 * Retell's v3 web calls don't pass the backend's `metadata` event or live
 * transcript updates through to the browser, so the backend publishes both to
 * Pusher instead (server/voice_events.py). Before dialing, the page makes up a
 * random id, subscribes to the public channel `voice-<id>`, and sends the id
 * in the call's metadata as `events_channel`.
 *
 * Neither value below is a secret. A key only lets you listen, and each call
 * listens on a channel named after an id only its own tab knows. */
export const PUSHER_KEY = process.env.NEXT_PUBLIC_PUSHER_KEY || "3b1aa13ad53a1a55ff03";
export const PUSHER_CLUSTER = process.env.NEXT_PUBLIC_PUSHER_CLUSTER || "us3";

/** Must match CHANNEL_PREFIX in server/voice_events.py. */
export const CHANNEL_PREFIX = "voice-";

/** How long a call waits for its channel (loading pusher-js included) before
 *  dialing anyway. The first event comes seconds into the call, so this
 *  rarely costs anything. */
export const SUBSCRIBE_TIMEOUT_MS = 3000;

export interface VoiceEventHandlers {
  onNavigation: (meta: NavigationMeta) => void;
  onTranscript: (window: VoiceLine[]) => void;
}

export interface VoiceEventsSubscription {
  /** Stop listening. Safe to call more than once. */
  close: () => void;
}

/** A fresh channel id: a lowercase UUID, the only shape the server accepts.
 *  crypto.randomUUID is missing on Safari before 15.4 and on plain-http pages
 *  (a phone trying a dev server over the LAN), so build one from
 *  getRandomValues there. */
export function newEventsChannelId(): string {
  if (typeof crypto.randomUUID === "function") return crypto.randomUUID();
  return uuidFromBytes(crypto.getRandomValues(new Uint8Array(16)));
}

/** A version 4 UUID from 16 random bytes. */
export function uuidFromBytes(bytes: Uint8Array): string {
  const b = Uint8Array.from(bytes);
  b[6] = (b[6] & 0x0f) | 0x40;
  b[8] = (b[8] & 0x3f) | 0x80;
  const hex = Array.from(b, (x) => x.toString(16).padStart(2, "0")).join("");
  return [hex.slice(0, 8), hex.slice(8, 12), hex.slice(12, 16), hex.slice(16, 20), hex.slice(20, 32)].join("-");
}

export function channelName(id: string): string {
  return CHANNEL_PREFIX + id;
}

/** The create-web-call body for a voice call whose events go to channel `id`.
 *  `events_channel` is the field server/voice_events.py reads. */
export function webCallBody(agentId: string, id: string, now = new Date()): CreateWebCallRequest {
  return {
    agent_id: agentId,
    metadata: {
      session_started: now.toISOString(),
      platform: "web",
      events_channel: id,
    },
  };
}

/** `promise`'s value, or null if it takes longer than `ms`. */
export function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T | null> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => resolve(null), ms);
    promise.then(
      (value) => {
        clearTimeout(timer);
        resolve(value);
      },
      (err) => {
        clearTimeout(timer);
        reject(err);
      },
    );
  });
}

/** A `navigation` event's payload, or null if it isn't one. */
export function parseNavigation(data: unknown): NavigationMeta | null {
  if (!data || typeof data !== "object") return null;
  const meta = data as Partial<NavigationMeta>;
  if (meta.type !== "navigation" || typeof meta.page !== "string") return null;
  if (meta.project_id !== undefined && typeof meta.project_id !== "string") return null;
  return meta as NavigationMeta;
}

/** A `transcript` event's lines, or null if there are none worth merging.
 *  Each needs its index in Retell's transcript, which is what the panel
 *  replaces lines by. */
export function parseTranscript(data: unknown): VoiceLine[] | null {
  if (!data || typeof data !== "object") return null;
  const lines = (data as { transcript?: unknown }).transcript;
  if (!Array.isArray(lines)) return null;
  const entries: VoiceLine[] = [];
  for (const line of lines) {
    if (!line || typeof line !== "object") continue;
    const { index, role, content } = line as { index?: unknown; role?: unknown; content?: unknown };
    if (
      Number.isInteger(index) &&
      (index as number) >= 0 &&
      (role === "agent" || role === "user") &&
      typeof content === "string" &&
      content
    ) {
      entries.push({ index: index as number, role, content });
    }
  }
  return entries.length > 0 ? entries : null;
}

const NOOP_SUBSCRIPTION: VoiceEventsSubscription = { close: () => {} };

/** Subscribe to a call's channel. Resolves once Pusher confirms the
 *  subscription, or once `timeoutMs` has passed since the call (loading
 *  pusher-js counts), whichever comes first. Never rejects: if Pusher can't
 *  be reached the call still works, the page just doesn't follow along. */
export async function subscribeToVoiceEvents(
  id: string,
  handlers: VoiceEventHandlers,
  timeoutMs = SUBSCRIBE_TIMEOUT_MS,
): Promise<VoiceEventsSubscription> {
  const deadline = Date.now() + timeoutMs;
  let Pusher: typeof import("pusher-js").default | null;
  try {
    // Loaded with the call, like the Retell SDK, so the page doesn't pay for
    // it. A chunk that stalls on a bad connection must not hold up the call.
    Pusher = await withTimeout(
      import("pusher-js").then((m) => m.default),
      timeoutMs,
    );
  } catch (err) {
    console.warn("Voice events unavailable:", err);
    return NOOP_SUBSCRIPTION;
  }
  if (!Pusher) {
    console.warn("Voice events unavailable: pusher-js took too long to load");
    return NOOP_SUBSCRIPTION;
  }

  const name = channelName(id);
  const pusher = new Pusher(PUSHER_KEY, { cluster: PUSHER_CLUSTER });
  const channel = pusher.subscribe(name);
  channel.bind("navigation", (data: unknown) => {
    const meta = parseNavigation(data);
    if (meta) handlers.onNavigation(meta);
  });
  channel.bind("transcript", (data: unknown) => {
    const window = parseTranscript(data);
    if (window) handlers.onTranscript(window);
  });

  let closed = false;
  const close = () => {
    if (closed) return;
    closed = true;
    // pusher-js keeps every instance it made; unbinding at least lets go of
    // the handlers, which hold this call's React setters.
    channel.unbind_all();
    pusher.unsubscribe(name);
    pusher.disconnect();
  };

  await new Promise<void>((resolve) => {
    const timer = setTimeout(resolve, Math.max(0, deadline - Date.now()));
    const done = () => {
      clearTimeout(timer);
      resolve();
    };
    channel.bind("pusher:subscription_succeeded", done);
    channel.bind("pusher:subscription_error", (err: unknown) => {
      console.warn("Voice events subscription failed:", err);
      done();
    });
  });
  return { close };
}
